genesiscommunitysuccess/custom-elements-lsp
# feat!: add plugin api and refactor fast functionality into its own plugin FUI-1331

⚠️ BREAKING CHANGE: The interface for adding FAST enhancements has changed ⚠️

🤔  &nbsp; **What does this PR do?**

`custom-elements-lsp`
- Expose an API which allows the loading of plugins.
- Add a plugin loader which loads any specified plugins
- Removes (moves) fast related source code (except passing it as an argument to the analyzer if required)
- BREAKING CHANGE:  tsconfig interface for enabling FAST has been moved to the `parser` options, and requires the plugin too

`cep-fast-plugin`
- New package which just contains all of the fast source files which were removed from `custom-elements-lsp`
- Source code files are the same except for the imports

`example`
- Implements the breaking change for the plugin interface

`example-lib` 
- Implements the breaking change for the plugin interface

Also updates readmes and the contributing document.

📑  &nbsp; **How should this be tested?**

Update test steps to match your PR:

```
1. Checkout branch
2. `npm run bootstrap`
3. `npm run test`

Other packages are completely tested via tests from `npm run test`. To test the LSP is working correctly in VSCode:
1. `cd packages/showcase/example`
2. `code .` (or manually open VSCode via the UI to the `example` directory on your filesystem).
3. Open a typescript file such as `root.ts`
4. In the Command Palette choose `TypeScript: select typescript version...` and then choose the workspace version
5. The LSP should be enabled. You will need to make a code change after the LSP is enabled before you see any Intellisense.
```

There is no new functionality from this PR, but the actions in the IDE should perform exactly the same as in master.

✅  &nbsp; **Checklist**
